### ISSUE TITLE
Visualizing performance characteristics of computations

### ISSUE THREAD

**Issue by jrbourbeau** [dask/dask#7141]
Users are sometimes caught off guard when they have a computation that works well and then add a new step which results in an overall computation with very different performance characteristics. For example, converting a CSV dataset to a parquet dataset

```python
import dask.dataframe ass dd

ddf = dd.read_csv(...)
ddf.to_parquet(...)
```

is easily paralizable. However, adding a `set_index` operation in the middle

```python
import dask.dataframe ass dd

ddf = dd.read_csv(...)
ddf = ddf.set_index(...)
ddf.to_parquet(...)
```

can impact performance significantly.

There are other operations like 

- `DataFrame.map_partitions` is easily paralizable
- `DataFrame.drop_duplicates` (and other methods which internally perform tree-reductions) which can produce large, interesting graph structures
- ...

It would be great if we had a visualization to communicate performance characteristics of various operations to users. For example, a graphviz-style plot at the `HighLevelGraph` `Layer` level where `set_index` layers are red and `map_paritions` layers are green. 

cc @jsignell @jacobtomlinson @rjzamora who may have an interest in thinking about visualizations

**Comment by mrocklin** [comment]
The streamz project does a good job of using both colors and shapes to denote different kinds of nodes.  

![image](https://user-images.githubusercontent.com/306380/106310280-26863980-6218-11eb-80e1-6b27ca8512df.png)

I think that we don't want to convey things as good vs bad, but instead we want to think about things like number of tasks, known streamability, etc..  And then from this information users should be able to answer questions like "should this stream well?" or others.

**Comment by jsignell** [comment]
What information do we have about the performance implications of different operations? Is this something that we already have catalogued, or would it be new information that we'd have to add? 

It would be potentially interesting to have a canonical dataset and just do different operations on that dataset and render the output of each `.visualize()`. It seems like you are talking about some kind of higher level visualization where the whole sections of the graph are collapsed and colored to denote something.

**Comment by mrocklin** [comment]
Correct.  I'm talking about a visualization on the high level graph, not the low level graph.  We don't have much information today except 

1. The number of tasks
2. The number of dependencies (with some work)

I think that in many cases we would annotate layers ourselves.  I don't expect this to be that hard in practice given subclassing.  Handling just blockwise and reductions for example would take care of a lot.

**Comment by jsignell** [comment]
>  a visualization on the high level graph

I think that could be super useful. Starting with the number of tasks seems like a natural place. I could imagine using size to denote that (or log of size).

**Comment by mrocklin** [comment]
cc @GenevieveBuckley

**Comment by mrocklin** [comment]
Adding to this a bit, I could imagine adding collection specific information to layers.  For arrays this might be chunks and dtype, for dataframes this might be partition structure, columns, and dtypes.  

Doing this is a little bit odd.  It adds information that is normally orthogonal onto the graph.  This breaks a bit of cleanliness.  It also might be an interesting step forward.  We might start using this strictly for visualization, and then later on decide that it's useful for other things, like optimization.  I don't anticipate this being deleterious in any way, assuming that when we serialize the graph that we don't bother sending this information just yet.  

This would allow us to provide much richer visualizations though, including things like column names and chunk structure for different stages in the graph.

**Comment by mrocklin** [comment]
If I were to do this I would probably add this information in the Array and Dataframe constructors

```python
class Array:
    def __init__(self, name, chunks, dtype, graph, ...):
        ...
        graph.layers[name].type = Array
        graph.layers[name].chunks = chunks
        graph.layers[name].dtype = dtype
```

(pseudocode, probably breaks many things) (also, same for dataframe)

I might do this, see if it passes tests, and then move on.

Next, I would probably start looking at what a high level visualization might be, and how we might build that visualization.  We currently have two different technologies for visualizing this information

1.  Basic HTML table views for nbytes, dtypes, etc..
2.  Our home-grown svg visualization code for chunk structure
3. `dot`/graphviz for high level graph layout

I don't know if it's easy to combine `dot` with custom svg images and tables, but if so, that would be grand.  If not then we might want to think about our own SVG layout for these graphs.  This would probably require some blackboard or pen-and-paper thinking.  


After this looks decent enough to be valuable then we might have enough motivation to include mangling the graph layers with collection metadata.  There would probably be an interesting conversation.

After that I would probably try to nerd-snipe others to use this same information for high level graph optimizations.  One possible ~sucker~ volunteer is @rjzamora, who has been thinking about pushing down column selection through the graph.

**Comment by GenevieveBuckley** [comment]
I can try this last suggestion in the next week or two and see what breaks. (Also happy for other people to jump in if that timeline doesn't work for you)

Once that's done, we'll be able to have a more informed discussion about the probable scope & best approaches to this.

**Comment by GenevieveBuckley** [comment]
Brief update:
I've added three lines to to dask Array `__new__` (I think this is what Matt was suggesting above)

```python
class Array(DaskMethodsMixin):
    ...
    def __new__(cls, dask, name, chunks, dtype=None, meta=None, shape=None):
        ...
        self.dask.layers[name].type = type(self)
        self.dask.layers[name].chunks = chunks
        self.dask.layers[name].dtype = dtype
        return self
```

Test failures look localized in two different areas.

### Failing tests from `test_array_core.py`
Running pytest, there are 7 failures in `pytest dask/array/tests/test_array_core.py -v`

The failures are almost all happening because `name` isn't a valid key in `self.dask.layers` for those particular tests. Usually, that's because we have `name = 'some-string'` and the key is a tuple like `('some-string', 0, 1)`. 

In one test (`test_constructor_plugin`), it says that `self` is actually a numpy array instead and gives us  `AttributeError: 'numpy.ndarray' object has no attribute 'dask'`.

### Failing tests from `test_distributed.py`
There are 14 failing tests if you run pytest on just this subset: `pytest dask/tests/test_distributed.py -v`

Oddly, if you run the whole pytest suite then you get 10 ERRORS from `test_distributed.py` but no actual test failures reported (maybe they're all skipped or something, idk).

I haven't looked at what's going wrong with these yet.

**Comment by GenevieveBuckley** [comment]
Thanks for writing out this wishlist @crusaderky

**Comment by GenevieveBuckley** [comment]
Update: https://github.com/dask/dask/pull/7309 is merged, which should help facilitate next steps for this issue.

**Comment by GenevieveBuckley** [comment]
@freyam I think you might  already subscribed to this issue, but if not, now you are :)

**Comment by mrocklin** [comment]
cc @seibert @eriknw @douglasdavis
